File: src/engine/platform.ts
import { invoke } from '@tauri-apps/api/core';
import { homeDir } from '@tauri-apps/api/path';

export interface PlatformInfo {
  os: 'macos' | 'linux' | 'windows';
  defaultShell: string;
  angyConfigDir: string;
}

let _cache: PlatformInfo | null = null;
let _promise: Promise<PlatformInfo> | null = null;

export async function getPlatformInfo(): Promise<PlatformInfo> {
  if (_cache) return _cache;
  if (!_promise) {
    _promise = invoke<PlatformInfo>('get_platform_info').then(info => {
      _cache = info;
      return info;
    }).catch(async () => {
      // Fallback: safe defaults using documented Tauri path API
      const home = await homeDir().catch(() => '~');
      _cache = {
        os: 'macos',
        defaultShell: '/bin/bash',
        angyConfigDir: `${home.replace(/\/+$/, '')}/.angy`,
      };
      return _cache!;
    });
  }
  return _promise;
}

export async function getModKey(): Promise<string> {
  const info = await getPlatformInfo();
  return info.os === 'macos' ? '⌘' : 'Ctrl+';
}

export async function getAngyConfigDir(): Promise<string> {
  const info = await getPlatformInfo();
  return info.angyConfigDir;
}

File: src-tauri/src/platform.rs
use serde::Serialize;
use tauri::command;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlatformInfo {
    pub os: String,
    pub default_shell: String,
    pub angy_config_dir: String,
}

#[command]
pub fn get_platform_info() -> Result<PlatformInfo, String> {
    let os = std::env::consts::OS.to_string(); // "linux", "macos", "windows"

    let default_shell = std::env::var("SHELL")
        .unwrap_or_else(|_| "/bin/bash".to_string());

    let home = std::env::var("HOME")
        .map_err(|_| "HOME not set".to_string())?;

    let angy_config_dir = if os == "linux" {
        let xdg_config = std::env::var("XDG_CONFIG_HOME")
            .unwrap_or_else(|_| format!("{}/.config", home));
        format!("{}/angy", xdg_config)
    } else {
        format!("{}/.angy", home)
    };

    Ok(PlatformInfo { os, default_shell, angy_config_dir })
}
